keps/sig-node/20190226-pod-overhead.md

```diff
@@ -0,0 +1,310 @@
+what is provided for the selected RuntimeClass, if one is specified.
+
+Kubelet's creation of the pod cgroup will be calculated as the sum of container
+`ResourceRequirements` fields, plus the Overhead associated with the pod.
+
+The scheduler, resource quota handling, and Kubelet's pod cgroup creation and eviction handling
+will take Overhead into account, as well as the sum of the pod's container requests.
+
+Horizontal and Veritical autoscaling are calculated based on container level statistics,
+so should not be impacted by pod Overhead.
+
+### API Design
+
+#### Pod overhead
+Introduce a Pod.Spec.Resources field on the pod to specify the pods overhead.
+
+```
+Pod {
+  Spec PodSpec {
+    // Overhead is the resource overhead consumed by the Pod, not including
+    // container resource usage. Users should leave this field unset.
+    // +optional
+    Overhead *ResourceRequirements
+  }
+}
+```
+
+For scheduling, the pod resource requests are added to the container resource requests.
+
+We don't currently enforce resource limits on the pod cgroup, but this becomes feasible once
+pod overhead is accountable. If the pod specifies a resource limit, and all containers in the
+pod specify a limit, then the sum of those limits becomes a pod-level limit, enforced through the
+pod cgroup.
+
+Users are not expected to manually set the pod resources; if a runtimeClass is being utilized,
+the manual value will be discarded. See RuntimeController for the proposal for setting these
+resources.
+
+Being able to specify resource requirements for a workload at the pod level instead of container level
+has been discussed, but isn't proposed in this KEP.
+
+In the event that pod-level requirements are introduced, pod overhead should be kept separate. This simplifies
+several scenarios:
+ - overhead, once added to the spec, stays with the workload, even if runtimeClass is redefined or removed.
+ - the pod spec can be referenced directly from scheduler, resourceQuote controller and kubelet, instead of referencing
+ a runtimeClass object which could have possibly been removed.
+
+
+### RuntimeClass changes
+
+Expand the runtimeClass type to include sandbox overhead, `Overhead *ResourceRequirements.`
+
+### RuntimeClass admission controller
+
+The pod resource overhead must be defined prior to scheduling, and we shouldn't make the user
+do it. To that end, we propose a new mutating admission controller: RuntimeClass. This admission controller
+is also proposed for the [native RuntimeClass scheduling KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md).
+
+In the scope of this KEP, The RuntimeClass controller will have a single job: set the pod overhead field in the
+workload's PodSpec according to the runtimeClass specified.
+
+It is expected that only the RuntimeClass controller will set Pod.Spec.Overhead. If a value is provided
```

Comment:
i think we should strip the overhead declaration from any pod that has no associated runtime class.